L1Trigger/TrackFindingTMTT/src/L1KalmanComb.cc

```diff
@@ -0,0 +1,1828 @@
+             << endl;
+#endif
+
+        if (getSettings()->kalmanDebugLevel() >= 4) {
+          cout << "dcov" << endl;
+          dcov.Print();
+          cout << "xcov" << endl;
+          state.last_state()->pxxa().Print();
+        }
+        TMatrixD h = H(stubCluster);
+        TMatrixD hxxh = HxxH(h, state.last_state()->pxxa());
+        if (getSettings()->kalmanDebugLevel() >= 4) {
+          cout << "h" << endl;
+          h.Print();
+          cout << "hxcovh" << endl;
+          hxxh.Print();
+        }
+        TMatrixD covR = dcov + hxxh;
+        if (getSettings()->kalmanDebugLevel() >= 4) {
+          cout << "covR" << endl;
+          covR.Print();
+          cout << "---" << endl;
+          cout << scientific << "delta = " << delta[0] << ", " << delta[1] << endl;
+        }
+        this->getDeltaChi2(covR, delta, false, deltaChi2rphi, deltaChi2rz);
+      }
+      chi2rphi = state.last_state()->chi2rphi() + deltaChi2rphi;
+      chi2rz = state.last_state()->chi2rz() + deltaChi2rz;
+#ifdef RECALC_DEBUG
+      cout << "  FITTER CHI2 UPDATE = " << (chi2rphi + chi2rz) << " delta chi2=" << (deltaChi2rphi + deltaChi2rz)
+           << " ID=" << ID << endl;
+#endif
+    }
+    return;
+  }
+
+  void L1KalmanComb::getDeltaChi2(const TMatrixD &dcov,
+                                  const std::vector<double> &delta,
+                                  bool debug,
+                                  double &deltaChi2rphi,
+                                  double &deltaChi2rz) const {
+    if (getSettings()->kalmanDebugLevel() >= 4) {
+      cout << "dcov" << endl;
+      dcov.Print();
+    }
+
+    if (dcov.Determinant() == 0) {
+      deltaChi2rphi = 999;
+      deltaChi2rz = 999;
+      return;
+    };
+
+    TMatrixD dcovi(dcov);
+    dcovi.Invert();
+
+    // Change in chi2 (with r-phi/r-z correlation term included in r-phi component)
+    deltaChi2rphi = delta.at(0) * delta.at(0) * dcovi(0, 0) + 2 * delta.at(0) * delta.at(1) * dcovi(0, 1);
+    deltaChi2rz = delta.at(1) * delta.at(1) * dcovi(1, 1);
+
+#ifdef RECALC_DEBUG
+    cout << "    FITTER DELTA CHI2: rphi=" << deltaCii2rphi << " rz=" << deltaChi2rz << endl;
+#endif
+
+    if (debug) {
+      cout << "CHI SQUARE OUTPUT" << endl;
+      cout << "cov" << endl;
+      dcov.Print();
+      cout << "cov inv" << endl;
+      dcovi.Print();
+      for (unsigned i = 0; i < delta.size(); i++)
+        cout << delta.at(i) << " ";
+      cout << endl;
+    }
+    return;
+  }
+
+  std::map<std::string, double> L1KalmanComb::getTrackParams(const L1KalmanComb *p, const KalmanState *state) {
+    return p->getTrackParams(state);
+  }
+
+  std::vector<double> L1KalmanComb::Hx(const TMatrixD &pH, const std::vector<double> &x) const {
+    std::vector<double> m((unsigned)pH.GetNrows(), 0);
+    if (pH.GetNcols() != (int)x.size()) {
+      cerr << "Hx() : H and x have different dimensions" << endl;
```

Comment:
Not thread safe, please use MessageLogger (or throw an exception?).

Reply:
done